### PR TITLE
osd/PG.cc: handle removal of pgmeta object

### PR DESCRIPTION
**Pull request by neha-ojha**
In 7f047005fc72e1f37a45cde2d742bb2eb1e62881, we made the pg removal code
much more efficient. But it started marking the pgmeta object as an unexpected
onode, which in reality is expected to be removed after all the other objects.

This behavior is very easily reproducible in a vstart cluster:

ceph osd pool create test 1 1
rados -p test bench 10 write --no-cleanup
ceph osd pool delete test test  --yes-i-really-really-mean-it

Before this patch:

"do_delete_work additional unexpected onode list (new onodes has appeared
since PG removal started[#2:00000000::::head#]" seen in the OSD logs.

After this patch:

"do_delete_work removing pgmeta object #2:00000000::::head#" is seen.

Related to:https://tracker.ceph.com/issues/50466
Signed-off-by: Neha Ojha <nojha@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dvanders** [comment]
@neha-ojha this will suppress the warning, but not solve the performance cost of the full collection list just above. Now that we understand the extra leftover object, can we just delete it directly instead of listing the entire collection?

File: src/osd/PG.cc [hunk 1]
```diff
@@ -2669,9 +2669,15 @@ std::pair<ghobject_t, bool> PG::do_delete_work(
       &olist,
       &next);
     if (!olist.empty()) {
-      dout(0) << __func__ << " additional unexpected onode list"
-              <<" (new onodes has appeared since PG removal started"
-              << olist << dendl;
+      for (auto& oid : olist) {
+        if (oid == pgmeta_oid) {
+          dout(20) << __func__ << " removing pgmeta object " << oid << dendl;
```
Comment (jdurgin) [inline src/osd/PG.cc:2674]: I think I'm missing something - where are we removing the pgmeta object?

**Comment by jdurgin** [comment]
> @neha-ojha this will suppress the warning, but not solve the performance cost of the full collection list just above. Now that we understand the extra leftover object, can we just delete it directly instead of listing the entire collection?

I think we're keeping the pgmeta object around so we can still open up the pg and continue if the osd crashes at this point - similar to not removing a directory until all files in it are gone

File: src/osd/PG.cc (hunk 1 above)
Comment (jdurgin) [reply src/osd/PG.cc:2674]: this is done by PGLog::clear_info_log() below

Comment (neha-ojha) [reply src/osd/PG.cc:2674]: It happens as a part of the final cleanup https://github.com/ceph/ceph/blob/master/src/osd/PG.cc#L2708-L2713, verified it in the logs

```
2021-04-22T21:42:28.416+0000 7fa19827e700 20 osd.0 pg_epoch: 16 pg[2.0( v 15'758 (0'0,15'758] lb MIN local-lis/les=13/14 n=758 ec=13/13 lis/c=13/13 les/c/f=14/14/0 sis=16 pruub=14.636423111s) [] r=-1 lpr=16 DELETING pi=[13,16)/1 crt=15'758 lcod 15'757 mlcod 0'0 unknown NOTIFY pruub 114.589828491s@ mbc={}] do_delete_work removing pgmeta object #2:00000000::::head#
2021-04-22T21:42:28.416+0000 7fa19827e700 10 bluestore(/home/nojha/work/ceph/build/dev/osd0) queue_transactions ch 0x565375807840 2.0_head
2021-04-22T21:42:28.416+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _txc_create osr 0x56537577d4a0 = 0x56537574f480 seq 791
2021-04-22T21:42:28.416+0000 7fa19827e700 20 _pin0x565367d68000   #2:00000000::::head# pinned
2021-04-22T21:42:28.416+0000 7fa19827e700 15 bluestore(/home/nojha/work/ceph/build/dev/osd0) _remove 2.0_head #2:00000000::::head# onode 0x565375901180 txc 0x56537574f480
2021-04-22T21:42:28.416+0000 7fa19827e700 15 bluestore(/home/nojha/work/ceph/build/dev/osd0) _do_truncate 2.0_head #2:00000000::::head# 0x0
2021-04-22T21:42:28.416+0000 7fa19827e700 20 bluestore.onode(0x565375901180).flush flush done
2021-04-22T21:42:28.420+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _do_omap_clear remove range start: 0x000000000000049F2D end: 0x000000000000049F7E
2021-04-22T21:42:28.420+0000 7fa19827e700 10 bluestore(/home/nojha/work/ceph/build/dev/osd0) _remove 2.0_head #2:00000000::::head# = 0
2021-04-22T21:42:28.420+0000 7fa19827e700 15 bluestore(/home/nojha/work/ceph/build/dev/osd0) _remove_collection 2.0_head
2021-04-22T21:42:28.420+0000 7fa19827e700 20 bluestore.OnodeSpace(0x5653758079f8 in 0x565367d68000) map_any
2021-04-22T21:42:28.420+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _collection_list range #-4:00000000::::0#0 to #-4:ffffffff::::0#0 and #2:00000000::::0#0 to #2:ffffffff::::0#0 start GHMIN
2021-04-22T21:42:28.420+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _collection_list pend #-4:ffffffff::::0#0
2021-04-22T21:42:28.420+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _collection_list oid #-1:30f0e826:::osdmap.13:0# >= #-4:ffffffff::::0#0
2021-04-22T21:42:28.424+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _collection_list oid #2:00000000::::head# end GHMAX
2021-04-22T21:42:28.424+0000 7fa19827e700 20 bluestore(/home/nojha/work/ceph/build/dev/osd0) _collection_list iterator not valid (end of db?)
2021-04-22T21:42:28.424+0000 7fa19827e700 10 bluestore(/home/nojha/work/ceph/build/dev/osd0) _remove_collection oid #2:00000000::::head#
```

**Review by jdurgin — APPROVED** [review]

**Review by xiexingguo — APPROVED** [review]
:+1:

File: src/osd/PG.cc [hunk 2]
```diff
@@ -2669,9 +2669,15 @@ std::pair<ghobject_t, bool> PG::do_delete_work(
       &olist,
       &next);
     if (!olist.empty()) {
-      dout(0) << __func__ << " additional unexpected onode list"
-              <<" (new onodes has appeared since PG removal started"
-              << olist << dendl;
+      for (auto& oid : olist) {
```
Comment (tchaikov) [inline src/osd/PG.cc:2672]: nit, i'd suggest drop the `if (!olist.empty)` check at line 2671.

**Comment by dvanders** [comment]
> > @neha-ojha this will suppress the warning, but not solve the performance cost of the full collection list just above. Now that we understand the extra leftover object, can we just delete it directly instead of listing the entire collection?
> 
> I think we're keeping the pgmeta object around so we can still open up the pg and continue if the osd crashes at this point - similar to not removing a directory until all files in it are gone

In that case, do we still need the entire block from line 2660 to 2682? AFAIU the unexpected leftover onode is now understood, is cleaned up elsewhere, so we can drop this expensive collection_list from the beginning.

**Comment by k0ste** [comment]
> @neha-ojha this will suppress the warning, but not solve the performance cost of the full collection list just above. Now that we understand the extra leftover object, can we just delete it directly instead of listing the entire collection?

I think @dvanders tells about latency spike and possibility to be marked down by mon. [Log](https://tracker.ceph.com/issues/50466#note-7) on tracker.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-05_13:07:31-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-07_01:47:01-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-07_12:44:46-rados-wip-kefu-testing2-2021-05-07-1516-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-08_03:28:06-rados-wip-kefu-testing2-2021-05-07-1516-distro-basic-smithi/

File: src/osd/PG.cc (hunk 2 above)
Comment (tchaikov) [reply src/osd/PG.cc:2672]: created #41233 to address this comment.

**Pull request merged**